pydantic/experimental/pipeline.py

```diff
@@ -0,0 +1,654 @@
+"""Experimental pipeline API functionality. Be careful with this API, it's subject to change."""
+
+from __future__ import annotations
+
+import datetime
+import operator
+import re
+from collections import deque
+from collections.abc import Container
+from dataclasses import dataclass
+from decimal import Decimal
+from functools import cached_property, partial
+from typing import TYPE_CHECKING, Any, Callable, Generic, Pattern, Protocol, TypeVar, Union, overload
+
+import annotated_types
+from typing_extensions import Annotated
+
+if TYPE_CHECKING:
+    from pydantic_core import core_schema as cs
+
+    from pydantic import GetCoreSchemaHandler
+
+from pydantic._internal._internal_dataclass import slots_true as _slots_true
+
+__all__ = ['validate_as', 'validate_as_deferred', 'transform']
+
+
+@dataclass(**_slots_true)
+class _ValidateAs:
+    tp: type[Any]
+    strict: bool = False
+
+
+@dataclass
+class _ValidateAsDefer:
+    func: Callable[[], type[Any]]
+
+    @cached_property
+    def tp(self) -> type[Any]:
+        return self.func()
+
+
+@dataclass(**_slots_true)
+class _Transform:
+    func: Callable[[Any], Any]
+
+
+@dataclass(**_slots_true)
+class _PipelineOr:
+    left: _Pipeline[Any, Any]
+    right: _Pipeline[Any, Any]
+
+
+@dataclass(**_slots_true)
+class _PipelineAnd:
+    left: _Pipeline[Any, Any]
+    right: _Pipeline[Any, Any]
+
+
+@dataclass(**_slots_true)
+class _Eq:
+    value: Any
+
+
+@dataclass(**_slots_true)
+class _NotEq:
+    value: Any
+
+
+@dataclass(**_slots_true)
+class _In:
+    values: Container[Any]
+
+
+@dataclass(**_slots_true)
+class _NotIn:
+    values: Container[Any]
+
+
+_ConstraintAnnotation = Union[
+    annotated_types.Le,
+    annotated_types.Ge,
+    annotated_types.Lt,
+    annotated_types.Gt,
+    annotated_types.Len,
+    annotated_types.MultipleOf,
+    annotated_types.Timezone,
+    annotated_types.Interval,
+    annotated_types.Predicate,
+    # common predicates not included in annotated_types
+    _Eq,
+    _NotEq,
+    _In,
+    _NotIn,
+    # regular expressions
+    Pattern[str],
+]
+
+
+@dataclass(**_slots_true)
+class _Constraint:
+    constraint: _ConstraintAnnotation
+
+
+_Step = Union[_ValidateAs, _ValidateAsDefer, _Transform, _PipelineOr, _PipelineAnd, _Constraint]
+
+_InT = TypeVar('_InT')
+_OutT = TypeVar('_OutT')
+_NewOutT = TypeVar('_NewOutT')
+
+
+class _FieldTypeMarker:
+    pass
+
+
+@dataclass(**_slots_true)
+class _Pipeline(Generic[_InT, _OutT]):
```

Comment:
I don't think we need to do it now, but I imagine we'll eventually want to make this type public, if only for type hinting purposes (e.g., being able to write a function that requires a pipeline with a given input type and output type as the input). At least, in the typescript world, I not-infrequently find myself using the equivalent of `_Pipeline[T, S]` in zod as a type hint.

Reply:
I guess also if we want to let users define functions that return the result of calling `validate_as` etc., possibly with some options, then they may want to annotate the return type of those functions using a public type.

Reply:
I've added a TODO in the code for this!